demo/app/controllers/demo_controller.rb

```diff
@@ -0,0 +1,151 @@
+        end
+      else
+        flash[:error] = @optimizely_service.errors
+      end
+    rescue StandardError => error
+      flash[:error] = error
+    end
+    render "new"
+  end
+
+  def visitors
+    # Returns list of visitors from model Visitor
+    @visitors = Visitor::VISITORS
+  end
+
+  def shop
+    # Calls before_action get_visitor from Application Controller to get visitor
+    # Calls before_action get_project_configuration from Private methods to get config object
+    # Calls before_action check_optimizely_client to check optimizely_client object exists
+    # Lists all products from Product model
+    # Calls optimizely client activate method to create variation(Static object) in OptimizelyService class
+    
+    @products = Product::PRODUCTS
+    @optimizely_service = OptimizelyService.new(@config.project_configuration_json)
+    if @optimizely_service.activate_service!(@visitor,@config.experiment_key )
+      @optimizely_service
+    else
+      flash[:error] = @optimizely_service.errors
+      redirect_to demo_config_path
+    end
+  end
+
+  def buy
+    # Calls before_action get_visitor from Application Controller to get visitor
+    # Calls before_action get_project_configuration from Private methods to get config object
+    # Calls before_action check_optimizely_client to check optimizely_client object exists
+    # Calls before_action get_product to get selected project
+    # Calls optmizely client's track method from OptimizelyService class
+    @optimizely_service = OptimizelyService.new(@config.project_configuration_json)
+    if @optimizely_service.track_service!(
+        @config.event_key,
+        @visitor,
+        @product.present? ? @product.except(:id) : {}
+    )
+      flash[:success] = "Successfully Purchased item #{@product[:name]} for visitor #{@visitor[:name]}!"
+    else
+      flash[:error] = @optimizely_service.errors
+    end
+    redirect_to messages_path
+  end
+
+  def log_messages
+    # Returns all log messages
+    @logs = LogMessage.all_logs
+  end
+
+  def delete_messgaes
+    LogMessage.delete_all_logs
+    redirect_to messages_path
+    flash[:success] = "log messages deleted successfully."
+  end
+
+  private
+
+  def demo_params
+    # Params passed on form submit to be permitted before save
+    params[:config].permit(:project_id, :experiment_key, :event_key,:project_configuration_json)
+  end
+
+  def validate_config!
+    @config = Config.find_or_create_by_project_id(demo_params[:project_id])
+    if @config.valid?
+      @config
+    else
+      flash[:error] = @config.errors.full_messages.first
+      render "new"
+    end
+  end
+
+  def get_project_configuration
+    @config = Config.find_by_project_id(session[:config_project_id])
+    unless @config.present? && @config.try(:experiment_key).present?
+      flash[:alert] = "Project id or Experiment key can't be blank!"
+      redirect_to demo_config_path
+    end
+  end
+
+  def get_or_generate_optimizely_client
```

Comment:
Please simplify logic for this method. Avoid nesting. Move else conditions to the top.

Reply:
Done!

Reply:
Code updated